modules/FvwmTalk/FvwmTalk.man
.\" t
.\" @(#)FvwmTalk.1	1/12/94
.TH FvwmTalk 1.20 "Jan 28 1994"
.UC
.SH NAME
FvwmTalk \- the FVWM command line interface
.SH SYNOPSIS
FvwmTalk is spawned by fvwm, so no command line invocation will work.

.SH DESCRIPTION
The FvwmTalk allows the user to type fvwm commands into a window, and
have them executed immediately. These commands are usually specfified
in the .fvwmrc file, or are bound to menu/mouse items as specified
in that file. This tools is particularly useful for testing new
configuration ideas, or for implementing temporary changes to your
environment.

.SH COPYRIGHTS
The FvwmTalk program, and the concept for
interfacing this module to the Window Manager, are all original work
by Robert Nation

Copyright 1994, Robert Nation. No guarantees or warranties or anything
are provided or implied in any way whatsoever. Use this program at your
own risk. Permission to use this program for any purpose is given,
as long as the copyright is kept intact.


.SH INITIALIZATION
So kill me, I can't remember what goes on.

.SH INVOCATION
FvwmTalk can be invoked by inserting the line 'Module FvwmTalk' in
the .fvwmrc file. This can be placed on a line by itself, if FvwmTalk
is to be spawned during fvwm's initialization, or can be bound to a
menu or mouse button or keystroke to invoke it later. Fvwm will search
directory specified in the ModulePath configuration option to attempt
to locate FvwmTalk.

.SH CONFIGURATION OPTIONS
I plead ignorance here. Read the code.


.SH AUTHOR
Robert Nation


ChangeLog
2026-02-23 David Flater

	Renamed package to Fvwm95 Forever and moved to GitHub.
	PACKAGE_VERSION = "Forever"

	Fixed FTBs for GCC 15.
	Updated things in configure.ac, Makefile.am, and system.fvwm95rc.in.
	Converted README to Markdown and revised as necessary.
	Merged BUGS and NEVERDO into README.md.

2020-12-10 David Flater

	fvwm95-2.0.43f-dwf8

	* Fixed multiple definitions of global variables in modules
	FvwmScroll and FvwmTaskBar.  GCC 10 changed the default behavior
	from -fcommon to -fno-common, which caused these multiple
	definitions to be rejected.

2020-03-18 David Flater

	fvwm95-2.0.43f-dwf7

	* Fixed buffer overflows in fvwm/read.c SendDataToModule which
	were triggered by a long IconPath or PixmapPath.

	* Added mini-firefox.xpm.

	* Updated system.fvwm95rc.in:
	- Added Firefox;
	- Added dosbox-x to the workaround for SDL2 fullscreen lossage;
	- Disabled Alt-arrow hotkeys (shortcuts to scroll by 1/10 of a
	page) to stop conflict with Doom (Alt = strafe).

	* Updated INSTALL.

2018-01-22 David Flater

	fvwm95-2.0.43f-dwf6

	* Updated system.fvwm95rc.in:
	- Fixed usage of killall;
	- Added the fullscreen workaround to chocolate-doom etc.;
	- Comments.

	* Updated INSTALL.

2017-04-20 David Flater

	fvwm95-2.0.43f-dwf5

	* Replaced faulty recursive build system with monolithic Automake:
        - Added script utils/makemakefileam.sh to generate Makefile.am;
	- Replaced configure.ac;
	- Purged all Imakefile and Makefile.in;
	- Adjusted install and search paths to agree with modern Linux.

	* Fixed innumerable instances of overflow, underflow, corruption,
	undefined behavior, division by zero, and broken parsing which
	were responsible for crashes and misbehavior in fvwm95 and every
	module.

	* Cleaned up compiler warnings through -Wall -Wno-parentheses on
	GCC 6.3.0.  Left the following:
	  ButtonArray.c:
	    warning: statement with no effect (BOGUS)
          etc:
  	    warning: the use of `mktemp' is dangerous, better use `mkstemp'
	    warning: implicit declaration of function 'ConsoleMessage' (BOGUS)

	* Harmonized swallow triggers in FvwmButtons, FvwmWharf,
	FvwmTaskBar, and FvwmScript (SwallowExec) to match on resource
	name (Resource), resource class (Class), or window name (Name),
	but not icon name (Icon Name).

	* Fixed fvwm95 bugs:
	- windows suddenly becoming enormous when resizing occurs near the
	  edge of the virtual desktop;
	- full-screen glitch when resizing a window that was declared with
  	  NoHandles and BorderWidth 0;
	- UseStyle option not accepting quoted identifiers;
	- false matching by UseStyle option;
	- no default WindowManagerName for Restart command (now fvwm95).

	* Fixed FvwmAudio bugs:
	- playing wrong or no sounds;
	- hanging, waiting on subprocesses to exit;
	- breaking audio_filename at whitespace.

	* Fixed FvwmConsole not using libreadline by default.

	* Fixed FvwmScript and sample scripts bugs:
	- missing default for FvwmScriptPath;
	- references to non-existent things in Buttons script;
	- infinite loop in PointerSetup script (but it is hopelessly fragile
	  and may hang on another system).

	* Fixed FvwmTaskBar bugs:
	- swallowing completely broken;
	- animation broken or too fast;
	- main event loop complexity;
	- no warning on missing *FvwmTaskBarGoodyLoadableSymbol;
	- bad geometry, location, and spacing of goodies;
	- bad calibration of x coordinates on mouse clicks on goodies;
	- area cleared different than area drawn;
	- missing default for FvwmTaskBarModuleIconPathBatStat;
	- only allowing one directory in FvwmTaskBarModuleIconPathBatStat.

	* Fixed FvwmWharf bugs:
	- black box over first button when swallowing failed;
	- wrong background colors on 32-bit arch;
	- animation too fast.

	* Corrected inaccuracies in man pages and HTML docs:
	- Documented the swallow commands for FvwmTaskBar;
	- Corrected description of the swallow commands for FvwmWharf;
	- Fixed ghost reference to FvwmInitBanner;
	- Corrected description of the name-matching behavior of the Style
  	  command to agree with what is actually done in fvwm/add_window.c
	  LookInList.

	* Made a few minor additions and enhancements:
	- Replaced the sample/default system.fvwm95rc;
	- Made FvwmTaskBar goodies work by single-click;
	- Improved error reporting by GoodyLoadable.c;
	- Added mini-xemacs.xpm, mini-chrome.xpm, mini-chromium.xpm,
	  mini-wilber16.xpm, and green.xpm;
	- Added LogoTest script.

	* Cleaned up random things:
	- Merged all needed preprocessor defines into FVWMconfig.h and made
	  all sources include it;
	- Removed version.h headers and declarations that conflicted with
	  FVWMconfig.h;
	- Merged index.html and GPL.html in with the other HTML files;
	- Merged xpmroot into utils;
	- Merged and purged duplicative icon collections:
	  - FvwmScript/Scripts/some_nice_icons:
	      moved shutdown.xpm to pixmaps/shutdown2.xpm
	      moved window.xpm to pixmaps/window2.xpm
	  - FvwmTaskBar/icons:
	      moved mini-mail.xpm to mini-icons/mini-mail2.xpm
	      moved minimail.xbm to mini-icons/minimail.xbm
	- Fixed saferealloc and made UpdateString use it (3 copies);
	- Moved random READMEs, example files, etc. from module
	  subdirectories into docs/modules;
	- Moved other random READMEs, example files, etc. into docs/random;
	- Moved the untested/unmaintained ASSound and OffiX subdirs of
	  FvwmWharf to docs/modules/FvwmWharf;
	- Deleted all other unused source files and scripts;
	- Normalized file permissions;
	- Stripped trailing whitespace.

	Not done:  see BUGS, NEVERDO.

2014-01-26 David Flater

	fvwm95-2.0.43f-dwf4

	* Fixed BatStatModule showing "charging" icon instead of "online"
	when fully charged.

	* Changed the "alert" and "empty" levels from 10% and 5% to 7% and
	2% respectively.

2014-01-25 David Flater

 	fvwm95-2.0.43f-dwf3

	* Fixed crash of FvwmTaskBar when a goody (BatStatModule) returned
	NULL from its init function to signify failure to init.

	* Updated BatStatModule to work with /sys/class/power_supply.

modules/FvwmTalk/FvwmTalk.c
#include <FVWMconfig.h>
/* This module, and the entire NoClutter program, and the concept for
 * interfacing this module to the Window Manager, are all original work
 * by Robert Nation
 *
 * Copyright 1994, Robert Nation. No guarantees or warantees or anything
 * are provided or implied in any way whatsoever. Use this program at your
 * own risk. Permission to use this program for any purpose is given,
 * as long as the copyright is kept intact. */

#define TRUE 1
#define FALSE

#define UPDATE_ONLY 1
#define ALL 2
#define PROP_SIZE 1024


#ifdef ISC
#include <sys/bsdtypes.h> /* Saul */
#endif

#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>

#if defined ___AIX || defined _AIX || defined __QNX__ || defined ___AIXV3 || defined AIXV3 || defined _SEQUENT_
#include <sys/select.h>
#endif

#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>
#include <X11/Xlib.h>
#include <X11/keysym.h>

#include <X11/Xutil.h>
#include <X11/Xproto.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>

#include "../../fvwm/module.h"

#include "FvwmTalk.h"

char *MyName;
int fd_width,screen, d_depth;
int fd[2];

int x_fd;
Window Root;
Display *dpy;
char *font_string = "helvetica";
char *ForeColor = "green";
char *BackColor = "black";
char *display_name = NULL;

XFontStruct *font;
Pixel fore_pix,back_pix;
static Atom wm_del_win;
unsigned long valuemask;
XSetWindowAttributes attributes;
XWMHints wmhints;
Window window;
char last_error[256],previous_line[256];
char Text[256];
int pos = 0;

XSizeHints sizehints =
{
  (PMinSize | PResizeInc | PBaseSize | PWinGravity | PMaxSize),
  0, 0, 100, 100,	                /* x, y, width and height */
  1, 1,		                        /* Min width and height */
  0, 0,		                        /* Max width and height */
  1, 1,	                         	/* Width and height increments */
  {0, 0}, {0, 0},                       /* Aspect ratio - not used */
  1, 1,                 		/* base size */
  (NorthWestGravity)                    /* gravity */
};

Pixel GetColor(char *name);
void nocolor(char *a, char *b);
XGCValues gcv;
GC myGC;
int My_XNextEvent(Display *dpy, XEvent *event);
void DrawWindow(int mode);
void paste_primary(int window,int property,int Delete);
void request_selection(int time);

/***********************************************************************
 *
 *  Procedure:
 *	main - start of module
 *
 ***********************************************************************/
int main(int argc, char **argv)
{
  char *temp, *s;
  XWMHints        wm_hints;
  XClassHint      class_hints;
  XTextProperty   window_name;

  /* Save the program name - its used for error messages and option parsing */
  temp = argv[0];

  s=strrchr(argv[0], '/');
  if (s != NULL)
    temp = s + 1;

  MyName = safemalloc(strlen(temp)+2);
  strcpy(MyName,"*");
  strcat(MyName, temp);

  if((argc != 6)&&(argc != 7))
    {
      fprintf(stderr,"%s Version %s should only be executed by fvwm!\n",MyName,
	      VERSION);
      exit(1);
    }

  /* Save the program name - its used for error messages and option parsing */
  temp = argv[0];

  s=strrchr(argv[0], '/');
  if (s != NULL)
    temp = s + 1;

  MyName = safemalloc(strlen(temp)+2);
  strcpy(MyName,"*");
  strcat(MyName, temp);

  /* Dead pipes mean fvwm died */
  signal (SIGPIPE, DeadPipe);

  fd[0] = atoi(argv[1]);
  fd[1] = atoi(argv[2]);

  /* Initialize X connection */
  if (!(dpy = XOpenDisplay(display_name)))
    {
      fprintf(stderr,"%s: can't open display %s", MyName,
	      XDisplayName(display_name));
      exit (1);
    }
  x_fd = XConnectionNumber(dpy);

  screen= DefaultScreen(dpy);
  Root = RootWindow(dpy, screen);
  if(Root == None)
    {
      fprintf(stderr,"%s: Screen %d is not valid ", MyName, (int)screen);
      exit(1);
    }
  d_depth = DefaultDepth(dpy, screen);

  fd_width = GetFdWidth();

  wm_del_win = XInternAtom(dpy,"WM_DELETE_WINDOW",False);

  /* load the font */
  if ((font = XLoadQueryFont(dpy, font_string)) == NULL)
    {
      if ((font = XLoadQueryFont(dpy, "fixed")) == NULL)
	{
	  fprintf(stderr,"%s: No fonts available\n",MyName);
	  exit(1);
	}
    };


  fore_pix = GetColor(ForeColor);
  back_pix = GetColor(BackColor);

  valuemask = (CWBackPixel | CWBorderPixel | CWEventMask);
  attributes.background_pixel = back_pix;
  attributes.border_pixel = fore_pix;
  attributes.event_mask = KeyPressMask | ExposureMask | ButtonPressMask;

  sizehints.width = 8*XTextWidth(font,"MMMMMMMMMM",10);
  sizehints.height = 3*(font->ascent+font->descent+2)+2;
  sizehints.x = 0;
  sizehints.y = 0;
  sizehints.width_inc = 0.1*XTextWidth(font,"MMMMMMMMMM",10);
  sizehints.height_inc = 3*(font->ascent+font->descent+2)+2;
  sizehints.base_width = 1;
  sizehints.base_height = 3*(font->ascent+font->descent+2)+2;
  sizehints.min_width = 1;
  sizehints.min_height = 3*(font->ascent+font->descent+2)+2;
  sizehints.max_height = 3*(font->ascent+font->descent+2)+2;
  sizehints.max_width = 26*XTextWidth(font,"MMMMMMMMMM",10);

  window = XCreateWindow (dpy, Root, 0, 0, sizehints.width,sizehints.height,
			  (unsigned int) 1,
			  CopyFromParent, InputOutput,
			  (Visual *) CopyFromParent,
			  valuemask, &attributes);

  class_hints.res_name = "FvwmTalk";
  class_hints.res_class =  "FvwmTalk";

  wm_hints.flags = InputHint;
  wm_hints.input = True;;

  XSetWMProtocols(dpy,window,&wm_del_win,1);
  XSetWMNormalHints(dpy,window,&sizehints);
  XStringListToTextProperty(&(argv[0]), 1, &window_name);
  XSetWMProperties(dpy, window, &window_name, &window_name,
      argv, argc, &sizehints, &wm_hints, &class_hints);

  gcv.foreground = fore_pix;
  gcv.background = back_pix;
  gcv.font = font->fid;
  myGC = XCreateGC(dpy,window,GCForeground|GCBackground|GCFont,&gcv);


  previous_line[0] = 0;
  last_error[0] = 0;
  XMapWindow(dpy,window);
  Loop(fd);

  return 0;
}


/***********************************************************************
 *
 *  Procedure:
 *	Loop - wait for data to process
 *
 ***********************************************************************/
void Loop(int *fd)
{
  KeySym keysym;
  static XComposeStatus compose = {NULL,0};
  XEvent event;
  char kbuf[100];
  int count;

  pos = 0;
  Text[0] = 0;

  while(1)
    {
      if(My_XNextEvent(dpy, &event))
	{
	  switch(event.type)
	    {
	    case KeyPress:
	      count = XLookupString(&(event.xkey),kbuf,99,&keysym,&compose);
	      kbuf[count] = (unsigned char)0;
	      if((keysym == XK_BackSpace )||(keysym == XK_Delete))
		{
		  if(pos > 0)
		    Text[--pos] = 0;
		}
	      else if((keysym == XK_Return)||(keysym == XK_KP_Enter))
		{
		  SendText(fd,Text,0);
		  last_error[0] = 0;
		  strncpy(previous_line,Text,255);
		  previous_line[255] = 0;
		  pos = 0;
		  Text[pos] = 0;
		  DrawWindow(ALL);
		}
	      else
		{
		  if(pos + count < 255)
		    {
		      strcat(Text,kbuf);
		      pos += count;
		    }
		  else
		    XBell(dpy,0);
		}
	      DrawWindow(UPDATE_ONLY);
	      break;
	    case ButtonPress:
	      if(event.xbutton.button == 2)
		request_selection(event.xbutton.time);
	      break;
	    case Expose:
	      DrawWindow(ALL);
	      break;
	    case ClientMessage:
	      if ((event.xclient.format==32) &&
		  (event.xclient.data.l[0]==wm_del_win))
		{
		  exit(0);
		}
	      break;
	    case SelectionNotify:
	      paste_primary(event.xselection.requestor,
			    event.xselection.property,True);
	    default:
	      break;
	    }
	}
    }
}


/***********************************************************************
 *
 *  Procedure:
 *	SIGPIPE handler - SIGPIPE means fvwm is dying
 *
 ***********************************************************************/
void DeadPipe(int nonsense)
{
  fprintf(stderr,"FvwmTalk: dead pipe\n");
  exit(0);
}



/****************************************************************************
 *
 * Loads a single color
 *
 ****************************************************************************/
Pixel GetColor(char *name)
{
  XColor color;
  XWindowAttributes attributes;

  XGetWindowAttributes(dpy, Root,&attributes);
  color.pixel = 0;
   if (!XParseColor (dpy, attributes.colormap, name, &color))
     {
       nocolor("parse",name);
     }
   else if(!XAllocColor (dpy, attributes.colormap, &color))
     {
       nocolor("alloc",name);
     }
  return color.pixel;
}


void nocolor(char *a, char *b)
{
  fprintf(stderr,"%s: can't %s %s\n", MyName, a,b);

}


/***************************************************************************
 *
 * Waits for next X event, or for an auto-raise timeout.
 *
 ****************************************************************************/
int My_XNextEvent(Display *dpy, XEvent *event)
{
  fd_set in_fdset;
  unsigned long header[HEADER_SIZE];
  static int miss_counter = 0;
  unsigned long *body;

  if(XPending(dpy))
    {
      XNextEvent(dpy,event);
      return 1;
    }

  FD_ZERO(&in_fdset);
  FD_SET(x_fd,&in_fdset);
  FD_SET(fd[1],&in_fdset);

#ifdef __hpux
  select(fd_width,(int *)&in_fdset, 0, 0, NULL);
#else
  select(fd_width,&in_fdset, 0, 0, NULL);
#endif

  if(FD_ISSET(x_fd, &in_fdset))
    {
      if(XPending(dpy))
	{
	  XNextEvent(dpy,event);
	  miss_counter = 0;
	  return 1;
	}
      else
	miss_counter++;
      if(miss_counter > 100)
	DeadPipe(0);
    }

  if(FD_ISSET(fd[1], &in_fdset))
    {
      if(ReadFvwmPacket(fd[1],header,&body) > 0)
	 {
	   if(header[1] == M_ERROR)
	     {
	       strncpy(last_error,(char *)(&body[3]),255);
	       last_error[strlen(last_error)-1] = 0;
	       last_error[255] = 0;
	       XClearArea(dpy,window,0,0,10000,10000,1);
	     }

	   free(body);
	 }
    }
  return 0;
}



void request_selection(int time)
{
  Atom sel_property;

  if (XGetSelectionOwner(dpy,XA_PRIMARY) == None)
    {
      /*  No primary selection so use the cut buffer.
       */
      paste_primary(DefaultRootWindow(dpy),XA_CUT_BUFFER0,False);
      return;
    }
  sel_property = XInternAtom(dpy,"VT_SELECTION",False);
  XConvertSelection(dpy,XA_PRIMARY,XA_STRING,sel_property,window,time);
}

void paste_primary(int window,int property,int Delete)
{
  Atom actual_type;
  int actual_format;
  unsigned long bytes_after, nitems, nread=0, i;
  unsigned char *data, *data2;

  if (property == None)
    return;

  do
    {
      if (XGetWindowProperty(dpy,window,property,nread/4,PROP_SIZE,Delete,
                             AnyPropertyType,&actual_type,&actual_format,
                             &nitems,&bytes_after,(unsigned char **)&data)
          != Success)
        return;
      if (actual_type != XA_STRING)
        return;

      data2 = data;
      /* want to make a \n to \r mapping for cut and paste only */
      for(i=0;i<nitems;i++)
        {
          if(*data == '\n')
            *data = '\r';
          data++;
        }

      if(255-pos > 0)
	{
	  strncat(Text,(char *)data2,255-pos-nitems);
	  pos = strlen(Text);
	  DrawWindow(UPDATE_ONLY);
	}
      nread += nitems;
      XFree(data2);
    } while (bytes_after > 0);
}

void DrawWindow(int mode)
{
  int w;

  if(mode == ALL)
    {
      XClearWindow(dpy,window);
      XDrawImageString(dpy,window,myGC,2,(font->ascent+font->descent+2)+
		       font->ascent+2,last_error,strlen(last_error));
      XDrawImageString(dpy,window,myGC,2, font->ascent+2,
		       previous_line,strlen(previous_line));
    }
  else
    {
      XClearArea(dpy,window,0,
		 2*(font->ascent+font->descent+2),10000,10000,0);
    }
  XDrawImageString(dpy,window,myGC,2,2*(font->ascent+font->descent+2)+
		   font->ascent+2,Text,pos);
  w=XTextWidth(font,Text,pos);
  XDrawLine(dpy,window,myGC,4+w,2*(font->ascent+font->descent+2)+2,
	    4+w,2*(font->ascent+font->descent+2)+
	    font->ascent+font->descent);
}
